### ISSUE TITLE
Check `!foo === bar` in `no-unsafe-negation` rule

### ISSUE THREAD

**Issue by fisker** [eslint/eslint#14802]
**What rule do you want to change?**

Forbid use `!foo === bar` and `!foo !== bar`

**Does this change cause the rule to produce more or fewer warnings?**

more

**How will the change be implemented? (New option, new default behavior, etc.)?**

Not sure

**Please provide some example code that this change will affect:**

```js
if (!foo === bar);
```

**What does the rule currently do for this code?**

Nothing

**What will the rule do after it's changed?**

Forbid it.

**Are you willing to submit a pull request to implement this change?**

Maybe.

---

I understand `!foo === bar` maybe just coverting falsely value to compare to a boolean, but most time it's a mistake,

I guess we should allow 

```js
// it's type-casting
if (!!foo === bar);
```

```js
if (!(foo === bar));
```

```js
if ((!foo) === bar);
```

I understand that stylistic rules are frozen, but still worth to make a proposal.
I can understand if you decide not to accept.

**Comment by bmish** [comment]
If this rule would catch "unsafe" behavior that "most of the time is a mistake" then it wouldn't be considered [stylistic](https://eslint.org/blog/2020/05/changes-to-rules-policies#what-are-the-changes) would it?

> Stylistic rules are those related to spacing, conventions, and generally anything that does not highlight an error or a better way to do something.

**Comment by coderaiser** [comment]
For [putout](https://github.com/coderaiser/putout) the rule will look like:

```js
module.exports.replace = () => ({
    '!__a === __b': '!__a',
    '!__a !== __b': '__a',
});
```

And it will end up with:

```js
// before
if (!foo === bar);
// after
if (!foo);
```

Because `bar` always `Boolean`. Or it can be:

```js
// before
if (!foo === bar);
// after
if (foo !== bar);
```

But if `bar` has type `Boolean` and `foo` has different type it will work in another way.
The runtime behavior will be changed in any case, but what kind of `autofix` you think is more correct?
Do you have examples of such code in your codebase? It looks like synthetic code with no examples in the wild, what kind of function could check this?

```js
function checkNot(x, y) {
    if (!x === y) {
    }
}

checkNot('hello', true);
```

This is very bad code, and autofix will broke it even more.

**Comment by fisker** [comment]
> but what kind of autofix you think is more correct?

Absolutely no auto-fix, this code is mistake.

> Do you have examples of such code in your codebase

https://github.com/prettier/prettier/commit/c7e354e38cfd051d3f64b56aeda35e9a0a669de6#r47946398

**Comment by coderaiser** [comment]
> > but what kind of autofix you think is more correct?
> 
> Absolutely no auto-fix, this code is mistake.
> 
> > Do you have examples of such code in your codebase
> 
> [prettier/prettier@c7e354e#r47946398](https://github.com/prettier/prettier/commit/c7e354e38cfd051d3f64b56aeda35e9a0a669de6#r47946398)

All right, such code:

```js
if (!left.type === "UnaryExpression" || left.operator !== "!") {}
```

Can and should be fixed to:

```js
if (left.type !== "UnaryExpression" || left.operator !== "!") {}
```

**Comment by fisker** [comment]
https://github.com/prettier/prettier/pull/10478/files

**Comment by coderaiser** [comment]
> Here is the PR fix that https://github.com/prettier/prettier/pull/10478/files

Just added your case to [@putout/plugin-simplify-ternary v1.1.0](https://github.com/coderaiser/putout/tree/master/packages/plugin-simplify-logical-expressions) 🎉

**Comment by mdjermanovic** [comment]
> Forbid use `!foo === bar` and `!foo !== bar`

Makes sense to me :+1: 

Thoughts about `==` and `!=`, too?

> **How will the change be implemented? (New option, new default behavior, etc.)?**
> 
> Not sure

This would have to be a new option.

> I guess we should allow
> 
> ```js
> // it's type-casting
> if (!!foo === bar);
> ```

Should we also allow code where the right side is obviously a boolean, eg. `!foo === true`?

> ```js
> if (!(foo === bar));
> ```     
>   
> ```js
> if ((!foo) === bar);
> ```

These would be [two suggestions](https://github.com/eslint/eslint/blob/39115c8b71d2629161359f6456f47fdbd552fddd/lib/rules/no-unsafe-negation.js#L103-L121) to fix the problem, like with other operators.

> I understand that stylistic rules are frozen, but still worth to make a proposal.

`no-unsafe-negation` reports possible errors, so it isn't a stylistic rule.

**Comment by fisker** [comment]
> Thoughts about == and !=, too?

I guess we can check them too.

> !foo === true

I prefer to report this case too, I believe this code is also a mistake most time. If it's intentionally to check boolean,

```js
const a = !foo === true;
const a = !foo !== false;
// ->
const a = !foo;

const b = !foo === false;
const b = !foo !== true;
// ->
const b = Boolean(foo);
```

**Comment by coderaiser** [comment]
> > Thoughts about == and !=, too?
> 
> I guess we can check them too.
> 
> > !foo === true
> 
> I prefer to report this case too, I believe this code is also a mistake most time. If it's intentionally to check boolean,
> 
> ```js
> const a = !foo === true;
> const a = !foo !== false;
> // ->
> const a = !foo;
> 
> const b = !foo === false;
> const b = !foo !== true;
> // ->
> const b = Boolean(foo);
> ```

By the way this case is handled by [@putout/plugin-remove-boolean-from-assertion](https://github.com/coderaiser/putout/tree/master/packages/plugin-remove-boolean-from-assertions).

**Comment by fisker** [comment]
@coderaiser Stop, please!

**Comment by nzakas** [comment]
@coderaiser you are starting to look a lot like spam at this point. Please stop.

**Comment by nzakas** [comment]
Looking at this more closely, I’m not sure adding checks for equality operators make sense. The point of this rule is to warn when negating a value will cause an unexpected result. It makes sense for the “in” and “instanceof” operators because you likely mistyped. It makes sense for comparison operators because you can’t really compare Boolean values. I’m having a hard time rationalizing that equality operators might be unsafe. This addition seems more like a stylistic preference than a warning about a potential problem because you can absolutely determine if a Boolean is equal to something else. 

So, I’m :-1: to adding this case to this rule.

**Comment by RastaAlex** [comment]
@coderaiser proposed solution how to solve the problem that he has already solved in his project.

**Comment by nzakas** [comment]
@btmills we could use your opinion on this one.

**Comment by btmills** [comment]
Checking `!foo === bar` seems in keeping with the spirit of this rule. The negation is _probably_ not what the author intended. There are valid uses of this patern: `!someInvertedValue === someBooleanVariable`, but I've seen intentional inverted comparisons to boolean variables far less than accidental instances of this pattern. Maybe it's a stylistic preference to say this is better written as `(!someInvertedValue) === someBooleanVariable`, but if this pattern is written unintentionally, it's a logic issue.

**Comment by mdjermanovic** [comment]
> Looking at this more closely, I’m not sure adding checks for equality operators make sense. The point of this rule is to warn when negating a value will cause an unexpected result. It makes sense for the “in” and “instanceof” operators because you likely mistyped. It makes sense for comparison operators because you can’t really compare Boolean values.

I agree with @nzakas on this. `!a in b` or `!a < b` almost never make sense, but `!a === b` is valid code if `b` is a boolean. Since we don't have type info, an error reported on such code would be a false positive, or could be interpreted as stylistic enforcement which isn't the intent of this rule.

I was in favor of this option in `no-unsafe-negation`, but now I think that we should report `!a === b` as a potential problem only if we're sure that the right side is not a boolean, so this rather belongs to `no-constant-binary-expression` rule (proposal https://github.com/eslint/eslint/issues/13752, there's `!foo == null` example in the original post), which will be based on determining types of the operands when possible. That would catch https://github.com/prettier/prettier/commit/c7e354e38cfd051d3f64b56aeda35e9a0a669de6#r47946398 as `<boolean> === <string>`.

**Comment by btmills** [comment]
> we should report !a === b as a potential problem only if we're sure that the right side is not a boolean, so this rather belongs to no-constant-binary-expression rule

That’s reasonable to me :+1:

**Comment by nzakas** [comment]
It seems like we can close this in favor of #13752?

**Comment by fisker** [comment]
I don't think the case is covered by that issue.
